### PR TITLE
Add cloudbuild configuration for automatic docker image builds

### PR DESCRIPTION
**Pull request by ajslone**
Please have a look. This adds a cloudbuild.json file that configures cloud build for our docker base images. This also adds a script to generate this file, and a configuration file for setting up our cloud builds. 
The config file in the PR has the url set to a test value for validation before we enable this for automatic builds.

### CODE REVIEW

File: cloudbuild.json [hunk 1]
```diff
@@ -0,0 +1,253 @@
+{
+  "steps": [
+    {
+      "id": "gcr.io/blueshift-playground/cloudbuild_test:gpu-base-100",
```
Comment (sritchie) [inline cloudbuild.json:4]: it'd be nice to explicitly say `cuda100` vs `100`, wdyt?

File: dockerfiles/Dockerfile [hunk 2]
```diff
@@ -44,9 +46,9 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
   git \
   python3 \
   python3-virtualenv \
-  wget \
-  && apt-get clean && \
```
Comment (sritchie) [inline dockerfiles/Dockerfile:48]: good catch!

File: dockerfiles/Dockerfile [hunk 3]
```diff
@@ -31,6 +31,8 @@ FROM $BASE_IMAGE
 MAINTAINER Sam Ritchie <sritchie09@gmail.com>
 
 ARG GCLOUD_LOC=/usr/local/gcloud
+ARG PYTHON_VERSION=3.8
```
Comment (sritchie) [inline dockerfiles/Dockerfile:34]: let's use 3.7 as the default. wdyt?

File: dockerfiles/Dockerfile [hunk 4]
```diff
@@ -31,6 +31,8 @@ FROM $BASE_IMAGE
...
+ARG MINICONDA_VERSION=py38_4.8.2
```
Comment (sritchie) [inline dockerfiles/Dockerfile:35]: what versions do they support? Is it every cross of python version and major-minor-incremental? Can you add a link here to the page that lists their versions? this sort of string hacking is tricky, I hear you.

File: scripts/cloudbuild.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+# ----------------------------------------------------------------------------
+logging.get_absl_handler().setFormatter(None)
+
+Config = Dict[str, Any]
+
+
+# ----------------------------------------------------------------------------
+class BuildStep(NamedTuple):
+  '''a google cloud-build step'''
+  sid: str
```
Comment (sritchie) [inline scripts/cloudbuild.py:38]: is this `step_id`? can we call it `id`, or `step_id`?

File: scripts/cloudbuild.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+  dockerfile: str = _CPU_DOCKERFILE
+  base_image: Optional[str] = _CPU_BASE_IMAGE
+  cpu_version: Optional[str] = None
+  gpu_version: Optional[str] = None
+  build_args: List[str] = []
+
+  def tag(self) -> str:
+    '''returns tag for this spec'''
+    parts = []
+    if self.gpu_version is None:  # cpu image
```
Comment (sritchie) [inline scripts/cloudbuild.py:65]: Can we pull this logic out into a separate function, and return some enum value? It is always easier for me to read these conditionals when they're simply defining a mode. Then we can switch-case on the mode inside here (and anywhere else that needs it).

File: scripts/cloudbuild.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        parts += ['base', self.gpu_version]
+      else:  # full gpu image
+        parts += [self.gpu_version, self.cpu_version]
+
+    return f'{FLAGS.base_url}:' + '-'.join(parts)
+
+
+# ----------------------------------------------------------------------------
+def _create_push_step(tag: str) -> BuildStep:
+  '''creates a cloudbuild push step'''
```
Comment (sritchie) [inline scripts/cloudbuild.py:79]: these docstrings aren't adding anything beyond the function name, which is perfectly descriptive. Delete?

File: scripts/cloudbuild.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+  return BuildStep(sid=f'{tag}-push', args=['push', tag], dependencies=[tag])
+
+
+# ----------------------------------------------------------------------------
+def _create_gpu_base_image_specs(cfg: Config) -> Dict[str, ImageSpec]:
+  '''creates gpu base image specs'''
```
Comment (sritchie) [inline scripts/cloudbuild.py:85]: Same here, etc

File: scripts/cloudbuild.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+def _create_image_spec(
+    cfg: Config,
+    cpu_specs: Dict[str, ImageSpec],
+    gpu_specs: Dict[str, ImageSpec],
+) -> ImageSpec:
+  '''creates an image spec from a config dictionary and gpu/gpu specs'''
+  gpu_version = cfg.get('gpu')
+  cpu_spec = cpu_specs[cfg.get('cpu')]
+
+  if gpu_version is None:
```
Comment (sritchie) [inline scripts/cloudbuild.py:110]: the mode switcher would help clean this up too.

File: scripts/cloudbuild_config.json [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+  "gpu_versions" : {
+    "100" : ["CUDA=10.0"],
+    "101" : ["CUDA=10.1"]
+  },
+  "cpu_versions" : {
+    "py37" : ["PYTHON_VERSION=3.7", "MINICONDA_VERSION=py37_4.8.2"],
+    "py38" : ["PYTHON_VERSION=3.8", "MINICONDA_VERSION=py38_4.8.2"]
+  },
+  "images" : [
+    {"cpu" : "py37"},
```
Comment (sritchie) [inline scripts/cloudbuild_config.json:11]: why is the python version a value, keyed by `cpu` here?

File: cloudbuild.json (hunk 1 above)
Comment (ajslone) [reply cloudbuild.json:4]: works for me, I'll change that

File: dockerfiles/Dockerfile (hunk 3 above)
Comment (ajslone) [reply dockerfiles/Dockerfile:34]: Fine by me, I'll change.

File: dockerfiles/Dockerfile (hunk 4 above)
Comment (ajslone) [reply dockerfiles/Dockerfile:35]: Not sure on this, looking at the conda docs, I would guess that they can support versions < that reflected in miniconda release tag, but that's just a guess. I'll add a link to the docs and the miniconda release archive here.

File: scripts/cloudbuild.py (hunk 5 above)
Comment (ajslone) [reply scripts/cloudbuild.py:38]: Agree, that's cleaner. I'll change that.

File: scripts/cloudbuild.py (hunk 7 above)
Comment (ajslone) [reply scripts/cloudbuild.py:79]: Agree, just feel such a compulsion to always have docstrings.

File: scripts/cloudbuild.py (hunk 8 above)
Comment (ajslone) [reply scripts/cloudbuild.py:85]: ack

File: scripts/cloudbuild_config.json (hunk 10 above)
Comment (ajslone) [reply scripts/cloudbuild_config.json:11]: Is the concern here that the 'cpu' key is misleading? Easy enough to rename this to 'python' if that helps.

File: scripts/cloudbuild.py (hunk 9 above)
Comment (ajslone) [reply scripts/cloudbuild.py:124]: Can't use that here, but modified other calls to use this mechanism.

File: dockerfiles/Dockerfile [hunk 11]
```diff
@@ -31,6 +31,12 @@ FROM $BASE_IMAGE
 MAINTAINER Sam Ritchie <sritchie09@gmail.com>
 
 ARG GCLOUD_LOC=/usr/local/gcloud
+ARG PYTHON_VERSION=3.7
+
+# minicoda release archive is here: https://repo.anaconda.com/miniconda
+# see the docs here for managing python versions with conda:
+# https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-python.html
+ARG MINICONDA_VERSION=py38_4.8.2
```
Comment (sritchie) [inline dockerfiles/Dockerfile:39]: needs to be py37

Comment (ajslone) [reply dockerfiles/Dockerfile:39]: Good catch, will fix this.

**Pull request merged**